Create CLI arg for assigning timeout value

chasen-bettinger opened the pull request:
Please fill in the fields below to submit a pull request.  The more information
that is provided, the better.

**Fixes issue #**:
https://github.com/in-toto/in-toto/issues/603

**Description of the changes being introduced by the pull request**:
    - Add `--run-timeout` CLI option to `in-toto-run` that enables user to
    specify a timeout in seconds for subprocess executions.
    - Modify function signature of `execute_link` to include `timeout` and
    function contents which uses that value in order to replace the previous
    behavior of only using the `LINK_CMD_EXEC_TIMEOUT` setting.


**Please verify and check that the pull request fulfills the following
requirements**:

- [x] The code follows the [Code Style Guidelines](https://github.com/secure-systems-lab/code-style-guidelines#code-style-guidelines)
- [x] Tests have been added for the bug fix or new feature
- [x] Docs have been added for the bug fix or new feature

adityasaky commented on in_toto/common_args.py:
```diff
@@ -195,6 +195,13 @@ def parse_password_and_prompt_args(args):
+RUN_TIMEOUT_KWARGS = {
+    "dest": "run_timeout",
+    "default": 10,
+    "help": ("float that represents the max timeout for the in-toto-run command"),
```
This should specify it's in seconds.

adityasaky replied:
And tbh, should we even support sub-second values? I don't know that we have use cases for this value that require such accuracy.

lukpueh replied:
+1 on full seconds

adityasaky commented on in_toto/runlib.py:
```diff
@@ -411,6 +410,7 @@ def in_toto_run(
     lstrip_paths=None,
     metadata_directory=None,
     use_dsse=False,
+    timeout=10
```
Rather than a magic number here, let's retain the default value in settings.py and use it here?

adityasaky commented on in_toto/runlib.py:
```diff
@@ -279,7 +279,7 @@ def _duplicate_streams():
 
 
-def execute_link(link_cmd_args, record_streams):
+def execute_link(link_cmd_args, record_streams, timeout=10):
```
We can drop the default value here as the expectation is this is private. AFAICT it's not documented on https://in-toto.readthedocs.io/en/latest/search.html?q=execute_link&check_keywords=yes&area=default

adityasaky commented on in_toto/runlib.py:
```diff
@@ -316,14 +315,14 @@ def execute_link(link_cmd_args, record_streams):
     if record_streams:
         return_code, stdout_str, stderr_str = _subprocess_run_duplicate_streams(
-            link_cmd_args, timeout=float(in_toto.settings.LINK_CMD_EXEC_TIMEOUT)
+            link_cmd_args, timeout=float(timeout)
```
On a similar note, this function shouldn't be changing the type IMO.

adityasaky reviewed (changes requested):
Good start, some initial thoughts.

lukpueh commented on tests/rc_test/.in_totorc:
```diff
@@ -1,5 +1,4 @@
 [settings]
 artifact_base_path=r/c/file
 ARTIFACT_EXCLUDE_PATTERNS=r:c:file
-LINK_CMD_EXEC_TIMEOUT=20
```
@adityasaky, looks like we forgot to remove this in f88138c90861953c77a1384ea2fcc58126e6fe59.

lukpueh replied:
@chasen-bettinger, you can revert this change and we'll remove the fixture in a separate PR.

adityasaky replied:
whoops, yeah

lukpueh commented on tests/test_runlib.py:
```diff
@@ -469,7 +469,8 @@ def test_record_follow_symlinked_directories(self):
             sorted(list(artifacts_dict.keys())),
-            sorted(self.full_file_path_list + [pair[1] for pair in link_pairs]),
+            sorted(self.full_file_path_list +
+                   [pair[1] for pair in link_pairs]),
```
This seems unrelated.

lukpueh replied:
... and breaks the black linter run. Do you maybe have a different max line length setting locally?

chasen-bettinger replied:
Think it was because I was using `autopep8` and not `black`.

lukpueh commented on tests/test_runlib.py:
```diff
@@ -1132,7 +1124,8 @@ def test_create_metadata_with_expected_product(self):
         link = Metablock.load(self.link_name)
-        self.assertEqual(list(link.signed.products.keys()), [self.test_product])
+        self.assertEqual(list(link.signed.products.keys()),
+                         [self.test_product])
```
This seems unrelated.

lukpueh replied:
see https://github.com/in-toto/in-toto/pull/605#discussion_r1211318961

lukpueh commented on doc/source/settings.rst:
```diff
@@ -7,4 +7,4 @@ globals directly in source code or at runtime. This method is discouraged. Inste
 
 .. note::
-  The globals ``DEBUG`` and ``LINK_CMD_EXEC_TIMEOUT`` can only be configured directly.
+  The globals ``DEBUG`` can only be configured directly.
```
```suggestion
  The global ``DEBUG`` can only be configured directly.
```

lukpueh commented on in_toto/common_args.py:
```diff
@@ -195,6 +195,13 @@ def parse_password_and_prompt_args(args):
+RUN_TIMEOUT_ARGS = ["--run-timeout"]
+RUN_TIMEOUT_KWARGS = {
+    "dest": "run_timeout",
+    "default": 10,
```
Let's also use the default from `settings.py` here.

lukpueh reviewed (commented):
Thanks for the contribution, this looks great! Please address @adityasaky's and my comments and we can merge this.

chasen-bettinger commented:
@adityasaky @lukpueh thanks for the review, updated PR accordingly!

adityasaky commented on in_toto/runlib.py:
```diff
@@ -316,14 +315,14 @@ def execute_link(link_cmd_args, record_streams):
             link_cmd_args,
             check=False,  # nosec
-            timeout=float(in_toto.settings.LINK_CMD_EXEC_TIMEOUT),
+            timeout=float(timeout),
```
```suggestion
            timeout=timeout,
```

adityasaky reviewed (commented):
This is looking pretty close! Some nits.

adityasaky commented:
@chasen-bettinger couple of other things:
* there's a mistaken import for the settings module
* missing DCO sign off on the last commit

Also, could you squash / generally clean up your commits? I think we're close to merging this one. :)

chasen-bettinger commented:
@adityasaky Take a look now.

adityasaky commented on in_toto/common_args.py:
```diff
@@ -195,6 +196,15 @@ def parse_password_and_prompt_args(args):
+    "dest": "run_timeout",
+    "default": LINK_CMD_EXEC_TIMEOUT,
+    "help": (
+        "integer that represents the max timeout in seconds for the in-toto-run command"
```
Can you include the default value in the help message?

adityasaky commented on in_toto/runlib.py:
```diff
@@ -475,6 +475,9 @@ def in_toto_run(
     use_dsse (optional): A boolean indicating if DSSE should be used to
         generate metadata.
 
+    timeout (optional): An integer indicating the max timeout in seconds 
```
Also include the default here.

adityasaky reviewed (approved):
LGTM, final nits from me. Can you amend your commit message? We're exposing the timeout to the user via the CLI and API in addition to the settings file. The intotorc file has been retired from the implementation minus the test artifact so it's not really involved in this PR.

adityasaky commented:
Also the linter is unhappy.

chasen-bettinger committed:
Create CLI arg for assigning timeout value

- Add `--run-timeout` CLI option to `in-toto-run` that enables user to
specify a timeout in seconds for subprocess executions.
- Modify function signature of `execute_link` to include `timeout` and
function contents which uses that value in order to replace the previous
behavior of only using the `LINK_CMD_EXEC_TIMEOUT` setting.

This commit closes #603.

Signed-off-by: Chasen Bettinger <bettingerchasen@gmail.com>

chasen-bettinger commented:
@adityasaky Ok, I think I've now made everyone happy.

lukpueh reviewed (approved):
Cheers!

adityasaky commented:
Thanks for you work on this @chasen-bettinger!
